org.uqbar.project.wollok.lib/src/wollok.wlk

```diff
@@ -246,15 +251,35 @@ package lang {
 		 * If more than one element satisfies the condition then it depends on the specific collection class which element
 		 * will be returned
 		 * @returns the element that complies the condition
+		 * @throws ElementNotFoundException if no element matched the given predicate
 		 * Example:
 		 *      users.find { user => user.name() == "Cosme Fulanito" }
 		 */
-		method find(predicate) = this.fold(null, { acc, e =>
-			 if (acc != null)
-			 	acc
-			 else
-			 	if (predicate.apply(e)) e else null
+		method find(predicate) = this.findOrElse(predicate, { 
+			throw new ElementNotFoundException("there is no element that satisfies the predicate")
 		})
+
+		/**
+		 * Returns the element of this collection that satisfy a given condition, or the given default otherwise, if no element matched the predicate
+		 * If more than one element satisfies the condition then it depends on the specific collection class which element
+		 * will be returned
+		 * @returns the element that complies the condition or the default value
+		 * Example:
+		 *      users.findOrElse({ user => user.name() == "Cosme Fulanito" }, homer)
+		 */
+		method findOrDefault(predicate, value) =  this.findOrElse(predicate, { value })
+		
+		/**
+		 * Returns the element of this collection that satisfy a given condition, 
+		 * or the the result of evaluating the given continuation 
+		 * If more than one element satisfies the condition then it depends on the specific collection class which element
+		 * will be returned
+		 * @returns the element that complies the condition or the result of evaluating the continuation
+		 * Example:
+		 *      users.findOrElse({ user => user.name() == "Cosme Fulanito" }, { homer })
+		 */
+		method findOrElse(predicate, continuation) native
```

Comment:
Is there any contraindication for the non local returns?
(I mean, besides the implementation complexity.)

On Fri, Jan 1, 2016 at 6:48 PM, Franco Leonardo Bulgarelli <
notifications@github.com> wrote:

> In org.uqbar.project.wollok.lib/src/wollok.wlk
> <https://github.com/uqbar-project/wollok/pull/565#discussion_r48680591>:
>
> > +		 * @returns the element that complies the condition or the default value
> > +		 * Example:
> > +		 *      users.findOrElse({ user => user.name() == "Cosme Fulanito" }, homer)
> > +		 */
> > +		method findOrDefault(predicate, value) =  this.findOrElse(predicate, { value })
> > +		
> > +		/**
> > +		 * Returns the element of this collection that satisfy a given condition,
> > +		 * or the the result of evaluating the given continuation
> > +		 * If more than one element satisfies the condition then it depends on the specific collection class which element
> > +		 * will be returned
> > +		 * @returns the element that complies the condition or the result of evaluating the continuation
> > +		 * Example:
> > +		 *      users.findOrElse({ user => user.name() == "Cosme Fulanito" }, { homer })
> > +		 */
> > +		method findOrElse(predicate, continuation) native
>
> PS: I am wondering that not having non-local returns AND no for statement
> AND no streams means means that there are many algorithms that can not be
> implemented directly in the language, right?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/uqbar-project/wollok/pull/565/files#r48680591>.
>